// @blacklab/react-image-magnify/dist/src/lib/imageRatio.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getContainerToImageRatio = exports.getLargeToSmallImageRatio = exports.getSmallToLargeImageRatio = void 0;
function getSmallToLargeElementRatio(smallElement, largeElement) {
    return {
        x: smallElement.width ? (largeElement.width / smallElement.width) : 1,
        y: smallElement.height ? (largeElement.height / smallElement.height) : 1,
    };
}
function getSmallToLargeImageRatio(smallImage, largeImage) {
    return getSmallToLargeElementRatio(smallImage, largeImage);
}
exports.getSmallToLargeImageRatio = getSmallToLargeImageRatio;
function getLargeToSmallImageRatio(smallImage, largeImage) {
    return {
        x: smallImage.width ? (smallImage.width / largeImage.width) : 1,
        y: smallImage.height ? (smallImage.height / largeImage.height) : 1,
    };
}
exports.getLargeToSmallImageRatio = getLargeToSmallImageRatio;
function getContainerToImageRatio(container, image) {
    return getSmallToLargeElementRatio(container, {
        width: image.width - container.width,
        height: image.height - container.height,
    });
}
exports.getContainerToImageRatio = getContainerToImageRatio;
